sklearn/linear_model/_glm/glm.py

```diff
@@ -423,27 +404,81 @@ def score(self, X, y, sample_weight=None):
+                "Some value(s) of y are out of the valid range of the loss"
+                f" {self._base_loss.__name__}."
+            )
+
+        # Note that constant_to_optimal_zero is already multiplied by sample_weight.
+        constant = np.mean(base_loss.constant_to_optimal_zero(y_true=y))
+        if sample_weight is not None:
+            constant *= sample_weight.shape[0] / np.sum(sample_weight)
+
+        # Missing factor of 2 in deviance cancels out.
+        deviance = base_loss(
+            y_true=y,
+            raw_prediction=raw_prediction,
+            sample_weight=sample_weight,
+            n_threads=1,
+        )
+        y_mean = base_loss.link.link(np.average(y, weights=sample_weight))
+        deviance_null = base_loss(
+            y_true=y,
+            raw_prediction=np.tile(y_mean, y.shape[0]),
+            sample_weight=sample_weight,
+            n_threads=1,
+        )
+        return 1 - (deviance + constant) / (deviance_null + constant)
 
     def _more_tags(self):
-        # create the _family_instance if fit wasn't called yet.
-        if hasattr(self, "_family_instance"):
-            _family_instance = self._family_instance
-        elif isinstance(self.family, ExponentialDispersionModel):
-            _family_instance = self.family
-        elif self.family in EDM_DISTRIBUTIONS:
-            _family_instance = EDM_DISTRIBUTIONS[self.family]()
+        # Create instance of BaseLoss if fit wasn't called yet. This is necessary as
+        # TweedieRegressor might set the used loss during fit different from
+        # self._base_loss.
+        base_loss = self._validate_base_loss()
+        return {"requires_positive_y": not base_loss.in_y_true_range(-1.0)}
+
+    def _validate_base_loss(self):
+        # This is only necessary because of the link and power arguments of the
+        # TweedieRegressor.
+        # Note that we do not need to pass sample_weight to the loss class as this is
+        # only needed to set loss.constant_hessian on which GLMs do not rely.
+        if not isinstance(self._base_loss, BaseLoss):
+            raise ValueError(
+                "The _base_loss must be an instance of sklearn._loss.loss.BaseLoss; "
+                f"got _base_loss_class={self._base_loss}"
+            )
+        return self._base_loss
+
+    # FIXME: remove in v1.3
+    @deprecated(  # type: ignore
+        "Attribute `family` was deprecated in version 1.1 and "
+        "will be removed in 1.3. Use `_base_loss` instead."
+    )
+    @property
+    def family(self):
+        if isinstance(self, PoissonRegressor):
+            return "poisson"
+        elif isinstance(self, GammaRegressor):
+            return "gamma"
+        elif isinstance(self, TweedieRegressor) and self.power == 3:
+            return "inverse-gaussian"
         else:
-            raise ValueError
-        return {"requires_positive_y": not _family_instance.in_y_range(-1.0)}
+            return self._base_loss.__class__.__name__
```

Comment:
I don't like deprecating without preserving the previous behavior. Either we delete this attribute entirely and document it as a breaking change or we keep it with the previous return values (not strings that are useless).

I have the feeling that it would be simpler to keep the `sklearn/_loss/_glm_distribution.py` file around during the deprecation cycle of the family attribute. We can just put a header at the top of the source file to make it explicit that this file is kept around until 1.3 just to preserve backward compat for the deprecated `family` parameter.